### ISSUE TITLE
Issue in inversion operation

### ISSUE THREAD

**Issue by varunsood189** [openssl/openssl#12692]
Hi, I am Varun. I am using openssl version openssl-1.1.1g build as static library on visual studio 2015.
While testing the Inversion, I got 0 as output.

For inversion: 
I am running following code:
Code:
********************************************************
```
int main()
{
       static BN_CTX *ctx = BN_CTX_new();
       //input       
       string n_string = "95A322A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFD5";       
       string b_string = "2A022A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFDD0";
       string result = "5F59561C4283177420AF0C0873C30BF44E0F574F5FE975619FCBDA2B9EAD6A88CD11224EC8BF10FB95A2C15A819DF91D4E0CD8DA7FCE3A6A51646CF52148D13C";
       
       BIGNUM *n = NULL, *a = NULL, *b = NULL, *ret = NULL;
       n = BN_new();
       a = BN_new();
       b = BN_new();
       ret = BN_new();
       BN_hex2bn(&n, n_string.c_str());  
       BN_hex2bn(&b, b_string.c_str());

       BIGNUM *assertion = NULL;
       assertion = BN_new();
       BN_gcd(assertion, b, n, ctx);                         // condition : gcd(e,lambda(n))==1
       if (!BN_is_one(assertion))
       {                          
             cout<< "Not invertible: N and b are not coprime. " << endl;
       }
       else {
             int field = 0;
             if (!field) //gf_p
             {
                    BIGNUM *inv = NULL;
                    inv = BN_new();
                    BN_mod_inverse(inv, b, n, ctx);
                    char *inv_ = BN_bn2hex(ret);
                    string inv_string(inv_);                 
                    cout << "Inverse result for GF(P)= " << inv_string << endl;
                    cout << "Correct result for GF(P)= " << result<< endl;
                    if (!strcmp(inv_, result.c_str()))
                    {
                           cout << " inversion is correct\n";
                    }
                    else {
                           cout << " inversion is not correct\n";
                    }
             }
             else { //gf_2m
                    BN_GF2m_mod_inv(ret, b, n, ctx);
                    char *inv_ = BN_bn2hex(ret);
                    string inv_string(inv_);                 
                    cout<< "Inverse result for GF(2^m)= " << inv_string << endl;
                    if (strcmp(inv_, result.c_str()))
                    {
                           cout << " inversion is correct\n";
                    }
                    else {
                           cout << " inversion is not correct\n";
                    }

             }
       }
       
       int rt;
       cin >> rt;
       BN_CTX_free(ctx);
       return 0;

}
```
********************************************************

Output:
********************************************************
Inverse result for GF(P)= 0
Correct result for GF(P)= 5F59561C4283177420AF0C0873C30BF44E0F574F5FE975619FCBDA2B9EAD6A88CD11224EC8BF10FB95A2C15A819DF91D4E0CD8DA7FCE3A6A51646CF52148D13C
inversion is not correct.

********************************************************
Is the testcase violating some condition in inversion? Am I making some mistake in above code ?

For Division:
Also in order to run the division, I am using combination of inversion and multiplication for mod gf(p).
I understand there is a function BN_div(), But as per my understanding, function is not valid for Galois Field GF(P).

**Comment by varunsood189** [comment]
Hi.
Is the issue valid?
Is more information required from me?
I would like to understand if i am doing anything wrong?
Is there any documentation on inversion implementation?
Regards,
Varun

**Comment by mattcaswell** [comment]
Perhaps @romen has a view?

**Comment by romen** [comment]
I fail to reproduce the issue in C with vanilla 1.1.1g in Linux.

I applied this patch on top of 1.1.1g and I get the expected results:

~~~diff
diff --git a/test/bntest.c b/test/bntest.c
index 97d08ac0be..673601bd9c 100644
--- a/test/bntest.c
+++ b/test/bntest.c
@@ -1471,6 +1471,30 @@ static int file_quotient(STANZA *s)
     return st;
 }
 
+static int file_modinv(STANZA *s)
+{
+    BIGNUM *a = NULL, *m = NULL, *mod_inv = NULL, *ret = NULL;
+    int st = 0;
+
+    if (!TEST_ptr(a = getBN(s, "A"))
+            || !TEST_ptr(m = getBN(s, "M"))
+            || !TEST_ptr(mod_inv = getBN(s, "ModInv"))
+            || !TEST_ptr(ret = BN_new()))
+        goto err;
+
+    if (!TEST_true(BN_mod_inverse(ret, a, m, ctx))
+            || !equalBN("A * x = 1 (mod M)", mod_inv, ret))
+        goto err;
+
+    st = 1;
+ err:
+    BN_free(a);
+    BN_free(m);
+    BN_free(mod_inv);
+    BN_free(ret);
+    return st;
+}
+
 static int file_modmul(STANZA *s)
 {
     BIGNUM *a = NULL, *b = NULL, *m = NULL, *mod_mul = NULL, *ret = NULL;
@@ -2696,6 +2720,7 @@ static int file_test_run(STANZA *s)
         {"Square", file_square},
         {"Product", file_product},
         {"Quotient", file_quotient},
+        {"ModInv", file_modinv},
         {"ModMul", file_modmul},
         {"ModExp", file_modexp},
         {"Exp", file_exp},
diff --git a/test/recipes/10-test_bn_data/bnmod.txt b/test/recipes/10-test_bn_data/bnmod.txt
index 5ea4d031f2..13a4dff3b3 100644
--- a/test/recipes/10-test_bn_data/bnmod.txt
+++ b/test/recipes/10-test_bn_data/bnmod.txt
@@ -5,10 +5,23 @@
 # in the file LICENSE in the source distribution or at
 # https://www.openssl.org/source/license.html
 
+# These test vectors satisfy A * ModInv = 1 (mod M) and 0 <= ModInv < M.
+
+Title = ModInv tests
+
+ModInv = 5F59561C4283177420AF0C0873C30BF44E0F574F5FE975619FCBDA2B9EAD6A88CD11224EC8BF10FB95A2C15A819DF91D4E0CD8DA7FCE3A6A51646CF52148D13C
+A = 2A022A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFDD0
+M = 95A322A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFD5
+
+
 # These test vectors satisfy A * B = ModMul (mod M) and 0 <= ModMul < M.
 
 Title = ModMul tests
 
+ModMul = 1
+A = 2A022A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFDD0
+B = 5F59561C4283177420AF0C0873C30BF44E0F574F5FE975619FCBDA2B9EAD6A88CD11224EC8BF10FB95A2C15A819DF91D4E0CD8DA7FCE3A6A51646CF52148D13C
+M = 95A322A68577F5491A462D72D611446DBC5BE09AF18BBAE8601B9A0ED8F84648B38281CD0DBC52841C12DCF0FC7A5EB2BE91AA71BE6AD215FF33421CAEEEDFD5
 
 ModMul = ae2ca2ce7addaee2e2b7752e286b2bb6a58b51cfbed5c924f00398e59ec36fe6341cd83da43a33a12410f45f6228079c4aeb3912be87e2e81fa1799151bfa0fea29873097475b2c3efa312145d0bf7e51b2a7c9bc961a4f4dcf0c883ff90b919b87c21099fba40257645be31f95a3a277
 A = 6b18497fed9befdf22a01d988d34213f6687d8a96e86c188dea4172e7c6095a0d18d3c86c0f5a1af9c6e3aaeb6baac2a510930b3ed06ec78ec2e12b
~~~

**Comment by romen** [comment]
@varunsood189 

~~~C
char *inv_ = BN_bn2hex(ret);
~~~

Should likely be

~~~C
char *inv_ = BN_bn2hex(inv);
~~~

**Comment by varunsood189** [comment]
Yes, I see the issue.  There is a coding issue. Thanks for reply. @romen 
I did not debug it because I was facing in Division(GF(P) and GF(2^m)) and Inversion (GF(P) and GF(2^m)) .
I am using the combination of inversion and multiplication to do division 
I also tried to understand the functions but i understand there is a function BN_div() , But as per my understanding, function is not valid for Galois Field GF(P).
```
	BIGNUM *inv = NULL;
	inv = BN_new();
	BN_mod_inverse(inv, b, n, ctx);
	BN_mod_mul(ret, a, inv, n, ctx);
	char *div_ = BN_bn2hex(ret);
	string div_string(div_);
	cout<< "Division result for GF(P)= " << div_string << endl;	

```
Inversion and division gf(p) are working. Thanks for the support.

**Comment by varunsood189** [comment]
But I am still facing some issue with inversion gf(2^m). And similarly with division gf(2^m).

```
int main()
{
static BN_CTX *ctx = BN_CTX_new();
//input
string n_string = "A04B1036278257C5FC20A1E35B15A2681374343CBBF2075FFA25681129467991";
string b_string = "0000020000000000000000000000000000000000000004000000000000000001";
string result = "167CF1E2D5046D3B40909F523877B25D25529083D22039761D2A267E3D06A098";

BIGNUM *n = NULL, *a = NULL, *b = NULL, *ret = NULL;
n = BN_new();
//a = BN_new();
b = BN_new();
ret = BN_new();
BN_hex2bn(&n, n_string.c_str());
//BN_hex2bn(&a, a_string.c_str());
BN_hex2bn(&b, b_string.c_str());

int field = 1;
if (field) //gf_2m
{	
	BN_GF2m_mod_inv(ret, b, n, ctx);
	char *inv_ = BN_bn2hex(ret);
	string inv_string(inv_);	
	cout << "Inverse result for GF(2^m)= " << inv_<< endl;	
	if (!strcmp(inv_, result.c_str()))
	{
		cout << " Inversion is correct\n";
	}
	else {
	cout << " Inversion is not correct\n";
	}
}

int rt;
cin >> rt;
BN_CTX_free(ctx);
return 0;
}

```

Every run I am getting different result.
Output:

Inverse result for GF(2^m)= 0FA5D117B037C242383697335E8E1BE4D5341C857F856B51D8D4E909A1A7BEE6
 Inversion is not correct

Inverse result for GF(2^m)= 5B84326B1AE47ABCFDEFD5DB283042027EE7A302E750625CB1000EAE7477001A
 Inversion is not correct

Is this behavior expected?

Regards,
Varun

**Comment by varunsood189** [comment]
Hi @romen  ,
Can you please try above usecase and let me know?
Thanks,
Varun

**Comment by romen** [comment]
@varunsood189 how are you generating the expected result?

**Comment by varunsood189** [comment]
Hi @romen ,

I referred it out of some old document. I am not 100% convinced it is correct.

If you have some reference for the same. Please let me know.

**Comment by romen** [comment]
There is currently one such test in our testsuite:

https://github.com/openssl/openssl/blob/09e76c5dd34515f9df42b2f1deed5166ba6b31fa/test/bntest.c#L771-L806

where `p0` and `p1` are defined as:

https://github.com/openssl/openssl/blob/09e76c5dd34515f9df42b2f1deed5166ba6b31fa/test/bntest.c#L48-L54

**Comment by romen** [comment]
For `GF(p)` usually I use SageMath to independently generate random values to play with and to use for testing against OpenSSL, but unfortunately I don't have any good similar suggestion for `GF(2m)`, as in SageMath they are far from trivial to handle!

**Comment by varunsood189** [comment]
Hi romen ,
I tried to use the multiplication operation in binary,
i got the following :

```
int main()
{
	int ret;
	static BN_CTX *ctx = BN_CTX_new();

	BIGNUM *n = NULL, *a = NULL, *b = NULL, *r = NULL;
	BIGNUM *inv = NULL;
	inv = BN_new();
	n = BN_new();
	a = BN_new();
	b = BN_new();
	r = BN_new();
	string n_string = "A04B1036278257C5FC20A1E35B15A2681374343CBBF2075FFA25681129467991";
	string b_string = "0000020000000000000000000000000000000000000004000000000000000001";
	string a_string = "167CF1E2D5046D3B40909F523877B25D25529083D22039761D2A267E3D06A098";
	cout << "n value = " << n_string << endl;
	cout << "b value = " << b_string << endl;
	cout << "a value = " << a_string << endl;
	
	BN_hex2bn(&n, n_string.c_str());
	BN_hex2bn(&b, b_string.c_str());
	BN_hex2bn(&a, a_string.c_str());
	BN_GF2m_mod_mul(r, a, b, n, ctx);
	cout << "r = "; BN_print_fp(stdout, r);  cout << endl;
        BN_CTX_free(ctx);
	cin >> ret;
	return ret;
}


```

i got the following output

n value = A04B1036278257C5FC20A1E35B15A2681374343CBBF2075FFA25681129467991
b value = 0000020000000000000000000000000000000000000004000000000000000001
a value = 167CF1E2D5046D3B40909F523877B25D25529083D22039761D2A267E3D06A098
r = 1

**Comment by romen** [comment]
@varunsood189 can you elaborate more on your last comment?

I am not sure how n, a, and b were generated and what are the expected properties of each. 

From a quick glance it would seem a and b are the inverse of each other modulo n.

I need more pointers to understand what seems to be the problem.

**Comment by varunsood189** [comment]
Hi , I was trying to use the inverse function in binary. `BN_GF2m_mod_inv` .
I doubted that the vectors have some problem in them so I tried out the multiplication 'BN_GF2m_mod_mul' in binary to check. 
Turns out a and b are inverse of each other.

I tried to find the reason behind the discrepancy, Binary Inverse implemented is using following:

> Modified Almost Inverse Algorithm (Algorithm 10) from Hankerson, D Hernandez, J.L., and Menezes, A.  "Software Implementation of Elliptic Curve Cryptography Over Binary Fields".

Could it be because the above vectors are not following some condition in the algorithm above. I am trying to understand are the test vectors valid.

**Comment by romen** [comment]
@varunsood189 I believe I am starting to see the issue you are seeing: using the values for `a`, `b`, and `n` from [your previous comment](https://github.com/openssl/openssl/issues/12692#issuecomment-744404494) you manage to compute `a * b (mod n)`, which returns `1` without errors, but if (using the same values) you try to compute `a^(-1) (mod n)` (expecting `b` as the result) you get an error:

~~~
n value = A04B1036278257C5FC20A1E35B15A2681374343CBBF2075FFA25681129467991
b value = 0000020000000000000000000000000000000000000004000000000000000001
a value = 167CF1E2D5046D3B40909F523877B25D25529083D22039761D2A267E3D06A098
r = 1
BN_GF2m_mod_inv failed
140190300628800:error:0308306A:bignum routines:BN_GF2m_mod:invalid length:crypto/bn/bn_gf2m.c:398:
~~~



<details>
<summary>Here is my C++ reproducer, derived from the code in <a href="https://github.com/openssl/openssl/issues/12692#issuecomment-744404494">your previous comment</a></summary>

~~~CPP
#include <iostream>
#include <string>
#include <assert.h>
#include <openssl/bn.h>
#include <openssl/err.h>

using namespace std;

int main()
{
    int ret = 1;
    BIGNUM *n = NULL, *a = NULL, *b = NULL, *r = NULL;
    BIGNUM *inv = NULL;
    BN_CTX *ctx = BN_CTX_new();

    string n_string = "A04B1036278257C5FC20A1E35B15A2681374343CBBF2075FFA25681129467991";
    string b_string = "0000020000000000000000000000000000000000000004000000000000000001";
    string a_string = "167CF1E2D5046D3B40909F523877B25D25529083D22039761D2A267E3D06A098";
    cout << "n value = " << n_string << endl;
    cout << "b value = " << b_string << endl;
    cout << "a value = " << a_string << endl;

    assert(ctx != NULL);
    inv = BN_CTX_get(ctx);
    n = BN_CTX_get(ctx);
    a = BN_CTX_get(ctx);
    b = BN_CTX_get(ctx);
    r = BN_CTX_get(ctx);
    assert(r != NULL);

    BN_hex2bn(&n, n_string.c_str());
    BN_hex2bn(&b, b_string.c_str());
    BN_hex2bn(&a, a_string.c_str());

    if (!BN_GF2m_mod_mul(r, a, b, n, ctx)) {
        cout << "BN_GF2m_mod_mul failed" << endl;
        goto end;
    }
    cout << "r = "; BN_print_fp(stdout, r);  cout << endl;

    if (!BN_GF2m_mod_inv(inv, a, n, ctx)) {
        cout << "BN_GF2m_mod_inv failed" << endl;
        goto end;
    }
    cout << "inv = "; BN_print_fp(stdout, r);  cout << endl;

    if (!BN_GF2m_mod_inv(inv, b, n, ctx)) {
        cout << "BN_GF2m_mod_inv failed" << endl;
        goto end;
    }
    cout << "inv = "; BN_print_fp(stdout, r);  cout << endl;


    cout << endl;

    ret = 0;
 end:
    if (ret != 0) ERR_print_errors_fp(stdout);
    BN_CTX_free(ctx);
    return ret;
}

~~~

</details>

The problem is that `n` is a "weird" polynomial, and that is why I asked where the values you are using came from.

It's "weird" because it has 117 non-zero terms, while usually we deal with trinomial or pentanomial bases.
I am not even sure it is a primitive polynomial and at the moment I don't have time to check it.

Anyway, the error is raised after returning from `BN_GF2m_poly2arr()` (which converts a BN_GF2m polynomial into its array representation) which in this backtrace is expecting at most 6 terms (pentanomial base plus an extra element as the array is terminated with a `-1`:
~~~
#0  BN_GF2m_poly2arr (a=0x55555556b418, p=0x7fffffffd050, max=6) at crypto/bn/bn_gf2m.c:1118
#1  0x00007ffff7d2d6a3 in BN_GF2m_mod (r=0x55555556b4c0, a=0x55555556b400, p=0x55555556b418) at crypto/bn/bn_gf2m.c:396
#2  0x00007ffff7d2e70d in BN_GF2m_mod_inv_vartime (r=0x55555556b400, a=0x55555556b400, p=0x55555556b418, ctx=0x55555556aeb0) at crypto/bn/bn_gf2m.c:570
#3  0x00007ffff7d2eded in BN_GF2m_mod_inv (r=0x55555556b400, a=0x55555556b430, p=0x55555556b418, ctx=0x55555556aeb0) at crypto/bn/bn_gf2m.c:745
#4  0x0000555555555778 in main ()
~~~

https://github.com/openssl/openssl/blob/90cebd1b216e0a160fcfd8e0eddca47dad47c183/crypto/bn/bn_gf2m.c#L1109-L1116

https://github.com/openssl/openssl/blob/90cebd1b216e0a160fcfd8e0eddca47dad47c183/crypto/bn/bn_gf2m.c#L384-L404

As you can see above, `BN_GF2m_mod()` is expecting `p` (your `n`) to be at most a pentanomial base (remember the extra array slot for the terminating `-1`), so line 398 is raising the error.

One could relax this limitation, at the cost of doing an extra `malloc` here instead of expecting at most 5 terms, which is why you don't see this error raised when calling `BN_GF2m_mod_mul()` with the same values.

 https://github.com/openssl/openssl/blob/90cebd1b216e0a160fcfd8e0eddca47dad47c183/crypto/bn/bn_gf2m.c#L458-L486

As you can see in `BN_GF2m_mod_mul()` the array is preallocated as having as many terms as there are bits in the `BIGNUM` representation (plus 1) of `p` (line 469).
And then at the end of the function it is freed.

This gives maximum flexibility, but the cost is very high, because for usual operations you would have at most 6 of those array elements actually in use. With your value of `n` you get to use `118` array elements, but still you are allocating `257` `int`s for it.
And it's not just the cost of "wasted" allocated but unused memory, but the problem that every `malloc` has usually a high cost: either a context switch to the operating system or a call to some userspace memory manager that has to find a big enough contiguous memory area for the mostly underutilized array.


## Conclusions / TL;DR

So, yes, there is an issue of `BN_GF2m_mod_inv()` (via `BN_GF2m_mod()`) being more strict than `BN_GF2m_mod_mul()` on the accepted values for the primitive polynomial given as input, but it does not apply to usual cryptographic applications and standards, which use at most pentanomial basis for GF2m.

If we want to resolve the inconsistency, I would argue that we might want to make `BN_GF2m_mod_mul()` more strict and efficient (i.e., remove the `malloc`/`free` and limit `arr`'s size to 6), rather than the other way around (i.e., changing `BN_GF2m_mod()` to be more flexible removing the "at-most-pentanomial" assumption and doing a `malloc` every time).

What does @openssl/otc think about this?


----

@varunsood189 does my answer satisfy your question about what condition the inputs do not satisfy?

I am quite busy and my `sagemath` installation at the moment requires maintenance, so I did not even try to play with your `a`, `b` and `n` there, but I would dare say that most likely that value of `n` is not coming from a set of test vectors for common GF2m cryptographic applications (unless it's something quite exotic or the output of some fuzzer trying random values).

I tried to see if `n` could have been a prime scalar, in case you grabbed it from some GFp test vector, but it is not prime:
13×19×730603×4128911×300042323827×324307969321951524597246101880249143773012057082953 (6 distinct prime factors)

So without knowing where these values came from it is difficult to build an argument for implementing changes to support arbitrary primitive polynomials for GF2m arithmetic in OpenSSL.
It doesn't seem `n`

**Comment by varunsood189** [comment]
Hi @romen , Thanks for the analysis above.

I was doubting the algorithm implementation but didn't get any error while running the algorithm, So was not able to get very far.

> I tried to see if n could have been a prime scalar, in case you grabbed it from some GFp test vector, but it is not prime:
> 13×19×730603×4128911×300042323827×324307969321951524597246101880249143773012057082953 (6 distinct prime factors
> 
I also thought the whether the N value is of GFp. 

That analysis has helped me understand the algortihm a little more. 

> I would argue that we might want to make BN_GF2m_mod_mul() more strict and efficient
I am not entirely sure making `BN_GF2m_mod_mul() ` is the best way forward. But its totally call of the team and following standards is the best way forward.

Thanks again for the above comment and effort you put in it.

**Comment by kroeckx** [comment]
> If we want to resolve the inconsistency, I would argue that we might want to make `BN_GF2m_mod_mul()` more strict and efficient (i.e., remove the `malloc`/`free` and limit `arr`'s size to 6), rather than the other way around (i.e., changing `BN_GF2m_mod()` to be more flexible removing the "at-most-pentanomial" assumption and doing a `malloc` every time).
> 
> What does @openssl/otc think about this?

So I think the question is if the functions can be used for general
math purpose or if we want to restrict it to only things used in
crypto. We're not a general math library, there are probably beter
projects for that. But I also don't see the need to restrict the
usage.

**Comment by romen** [comment]
So we have 3 options:
A) leave things as they are, with the inconsistency on the input conditions for these two affine operations. The inconsistency only happens for unusual inputs outside the scope of openssl cryptographic applications
B) we solve the inconsistency relaxing the modular inverse at the cost of performance and memory usage
C) we solve the inconsistency making modular multiplication stricter, improving performance and memory usage for its users at the potential cost of breaking applications that were using OpenSSL as a generic math library with unusual inputs for cryptographic applications

**Comment by t8m** [comment]
I vote for A. At least for OpenSSL 3.0

**Comment by varunsood189** [comment]
> A) leave things as they are, with the inconsistency on the input conditions for these two affine operations. The inconsistency only happens for unusual inputs outside the scope of openssl cryptographic applications

Leaving as these are will not solve above problem but as you said if its outside the scope of openssl. I think their should be a disclaimer or some info on it in documentation. right?

> B) we solve the inconsistency relaxing the modular inverse at the cost of performance and memory usage

I would prefer it. Since algorithm does not restrict input vectors.

> C) we solve the inconsistency making modular multiplication stricter, improving performance and memory usage for its users at the potential cost of breaking applications that were using OpenSSL as a generic math library with unusual inputs for cryptographic applications

Breaking something is bad some system might be affected because of it. Is there any algorithm which uses `BN_GF2m_mod_mul()` or `BN_GF2m_mod_inv() ` operations inside the algorithm to perform inversion and division. I have not seen any algorithm yet. Which uses these operations.

How do we decide? is there a panel  which decides or voting?

**Comment by romen** [comment]
<https://ia.cr/2007/192> (and section 2 in particular) could be useful as a quick reminder of why trinomial or pentanomial basis for gf2m are sufficient and preferred for efficient cryptographic applications.

@varunsood189 also has a point on the fact that we are lacking documentation for these functions (even the public ones), as we are missing pod files for them.

@varunsood189 would you be willing to help us on this aspect and add the missing documentation files? (including whatever conditions we are going to decide upon)

Luckily most of these functions have documentation in the form of comments next to the function definition, so it would be initially a matter of using those as the source of the novel pod files.
During this process you might want to add a few tests to check that what is documented reflects what is actually implemented.

What do you say? 


For the decision process it depends: we can reach an agreement here in this discussion and bring it to the attention of the OTC or OMC if that is required.

**Comment by varunsood189** [comment]
Hi @romen,

I don't mind helping out to increase the documentation whenever I get some free time. But I should probably mention my knowledge in GFp and GF2m is still limited. I might need help from you or some other members.
Also, What are pod files?

**Comment by romen** [comment]
Yes, of course we can help with revising and guiding!

When I mentioned "pod files" I was referring to files in the [Plain Old Documentation](https://en.wikipedia.org/wiki/Plain_Old_Documentation) format: for most public functions we have an entry in `doc/man3/<function_name>.pod`.

Acutally it is a little bit more tricky as a single `.pod` file might list several functions within, and then during the build stage it is the build system that takes care of creating duplicates of the resulting manpages or htmlpages (that is what we compile those `.pod` files to) so that for every listed function you have a copy/symlink/hardlink with the name of the function and the contents of the `.pod` file.

For example if you visit https://www.openssl.org/docs/manmaster/man3/BN_sub.html you will see the same contents as visiting `BN_add.html` (but also of `BN_mul`, `BN_sqr`, `BN_div`, `BN_mod`, `BN_nnmod`, `BN_mod_add`, `BN_mod_sub`, `BN_mod_mul`, `BN_mod_sqr`, `BN_exp`, `BN_mod_exp`, `BN_gcd`).
All these files are generated from [`doc/man3/BN_add.pod`](https://raw.githubusercontent.com/openssl/openssl/master/doc/man3/BN_add.pod) which at the top, in the `NAME` section, lists all the functions documented within that `.pod` file and for which the build system will generate properly named duplicates.

**Comment by varunsood189** [comment]
Hi Can you try to build following code as .pod

```
=pod

=head1 NAME

BN_GF2m_add ,BN_GF2m_sub ,BN_GF2m_mod ,BN_GF2m_mod_mul ,BN_GF2m_mod_sqr ,BN_GF2m_mod_inv ,BN_GF2m_mod_div ,BN_GF2m_mod_exp ,BN_GF2m_mod_sqrt ,BN_GF2m_mod_solve_quad ,BN_GF2m_cmpa ,BN_GF2m_mod_arr, BN_GF2m_mod_mul_arr ,BN_GF2m_mod_sqr_arr ,BN_GF2m_mod_inv_arr ,BN_GF2m_mod_div_arr ,BN_GF2m_mod_exp_arr ,BN_GF2m_mod_sqrt_arr ,BN_GF2m_mod_solve_quad_arr ,BN_GF2m_poly2arr ,BN_GF2m_arr2poly -
arithmetic operations on binary BIGNUMs

=head1 SYNOPSIS

 #include <openssl/bn.h>

int BN_GF2m_add(BIGNUM *r, const BIGNUM *a, const BIGNUM *b);

int BN_GF2m_sub(BIGNUM *r, const BIGNUM *a, const BIGNUM *b);

int BN_GF2m_mod(BIGNUM *r, const BIGNUM *a, const BIGNUM *p);

int BN_GF2m_mod_mul(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_sqr(BIGNUM *r, const BIGNUM *a, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_inv(BIGNUM *r, const BIGNUM *b, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_div(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_exp(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_sqrt(BIGNUM *r, const BIGNUM *a, const BIGNUM *p, BN_CTX *ctx);

int BN_GF2m_mod_solve_quad(BIGNUM *r, const BIGNUM *a, const BIGNUM *p, BN_CTX *ctx);

# define BN_GF2m_cmp(a, b) BN_ucmp((a), (b))

int BN_GF2m_mod_arr(BIGNUM *r, const BIGNUM *a, const int p[]);

int BN_GF2m_mod_mul_arr(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_sqr_arr(BIGNUM *r, const BIGNUM *a, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_inv_arr(BIGNUM *r, const BIGNUM *b, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_div_arr(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_exp_arr(BIGNUM *r, const BIGNUM *a, const BIGNUM *b, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_sqrt_arr(BIGNUM *r, const BIGNUM *a, const int p[], BN_CTX *ctx);

int BN_GF2m_mod_solve_quad_arr(BIGNUM *r, const BIGNUM *a, const int p[], BN_CTX *ctx);

int BN_GF2m_poly2arr(const BIGNUM *a, int p[], int max);

int BN_GF2m_arr2poly(const int p[], BIGNUM *a);

=head1 DESCRIPTION

BN_GF2m_add() Add polynomials I<a> and I<b> and store result in I<r> (C<r = a + b>). I<r> could be I<a> or I<b>, I<a> and I<b> could be equal I<r> is the bitwise XOR of I<a> and I<b>.

BN_GF2m_sub() Sub polynomials I<a> and I<b> and store result in I<r>; Same as BN_GF2m_add() 

BN_GF2m_mod() Performs modular reduction of I<a> by I<p> and store result in I<r> (C< r=a mod p>). I<r> could be I<a>. This function calls down to the BN_GF2m_mod_arr implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_arr function.

BN_GF2m_mod_mul() Compute the product of two polynomials I<a> and I<b>, reduce modulo I<p>, and store the result in I<r> (C< r = (a * b) mod p>). I<r> could be I<a> or I<b> I<a> could equal I<b>. This function calls down to the BN_GF2m_mod_mul_arr implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_mul_arr function.

BN_GF2m_mod_sqr() Square I<a>, reduce the result mod I<p>, and store it in I<a>. I<r> could be I<a> (C< r = (a * a) mod p>). This function calls down to the BN_GF2m_mod_sqr_arr implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_sqr_arr function.

BN_GF2m_mod_inv() Inverse of polynomials and I<b>, reduce modulo I<p>, and store the result in I<r> (C< r = (a / b) mod p>). Wrapper for BN_GF2m_mod_inv_vartime that blinds the input before calling. This is not constant time. But it does eliminate first order deduction on the input.

BN_GF2m_mod_div() Divide I<y> by I<x>, reduce modulo I<p>, and store the result in I<r> (C< r = (1 / b) mod p >). I<r> could be I<x> or I<y>, I<x> could equal I<y>.

BN_GF2m_mod_exp() Compute the bth power of I<a>, reduce modulo I<p>, and store the result in I<r> (C< r = (a ^ b) mod p >). I<r> could be I<a>. This function calls down to the BN_GF2m_mod_exp_arr implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_exp_arr function.

BN_GF2m_mod_sqrt() Compute the square root of I<a>, reduce modulo I<p>, and store the result in I<r> (C< r = sqrt(a) mod p >) . I<r> could be I<a>. This function calls down to the BN_GF2m_mod_sqrt_arr implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_sqrt_arr function.

BN_GF2m_mod_solve_quad() Find I<r> such that (C<r^2 + r = a mod p>). I<r> could be I<a>. If no I<r> exists returns 0. This function calls down to the BN_GF2m_mod_solve_quad_arr implementation; this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_solve_quad_arr function.

BN_GF2m_cmp() Same as BN_ucmp() 

BN_GF2m_mod_arr() Performs modular reduction of I<a> and store result in I<r>(C< r = a mod p >). I<r> could be I<a>.

BN_GF2m_mod_mul_arr() Compute the product of two polynomials I<a> and I<b>, reduce modulo I<p>, and store the result in I<r> (C< r = (a * b) mod p >). I<r> could be I<a> or I<b>; I<a> could be I<b>.

BN_GF2m_mod_sqr_arr() Square I<a>, reduce the result mod I<p>, and store it in I<a>. I<r> could be I<a> (C< r = (a * a) mod p >)

BN_GF2m_mod_inv_arr() Invert I<b>, reduce modulo I<p>, and store the result in I<r> (C< r = (1 / b) mod p >). I<r> could be I<b>. This function calls down to the BN_GF2m_mod_inv implementation this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_inv function.

BN_GF2m_mod_div_arr() Divide I<a> by I<b>, reduce modulo I<p>, and store the result in I<r> (C<  r = (a / b) mod p  >). I<r> could be I<b> or I<a>, I<b> could equal I<a>. This function calls down to the BN_GF2m_mod_div implementation; this wrapper function is only provided for convenience for best performance, use the BN_GF2m_mod_div function.

BN_GF2m_mod_exp_arr() Compute the bth power of I<a>, reduce modulo I<p>, and store the result in I<r>(C< r = (a ^ b) mod p >). I<r> could be I<a>. Uses simple square-and-multiply algorithm A.5.1 from IEEE P1363.

BN_GF2m_mod_sqrt_arr() Compute the square root of I<a>, reduce modulo I<p>, and store the result in I<r>(C< r = sqrt(a) mod p >). I<r> could be I<a>. Uses exponentiation as in algorithm A.4.1 from IEEE P1363.

BN_GF2m_mod_solve_quad_arr() Find I<r> such that (C<r^2 + r = a mod p>). I<r> could be I<a>. If no I<r> exists returns 0. Uses algorithms A.4.7 and A.4.6 from IEEE P1363.

BN_GF2m_poly2arr() Convert the bit-string representation of a polynomial (C<( \sum_{i=0}^n a_i * x^i)>) into an array of integers corresponding to the bits with non-zero coefficient. Array is terminated with -1. Up to max elements of the array will be filled. Return value is total number of array elements that would be filled if array was large enough.

BN_GF2m_arr2poly() Convert the coefficient array representation of a polynomial to I<a> bit-string. The array must be terminated by -1.

For all functions, I<ctx> is a previously allocated B<BN_CTX> used for
temporary variables; see L<BN_CTX_new(3)>.

Unless noted otherwise, the result B<BIGNUM> must be different from
the arguments.

=head1 RETURN VALUES

For all functions, 1 is returned for success, 0 on error. The return
value should always be checked (e.g., C<if (!BN_GF2m_add(r,a,b)) goto err;>).
The error codes can be obtained by L<ERR_get_error(3)>.

=head1 SEE ALSO

L<ERR_get_error(3)>, L<BN_CTX_new(3)>,
L<BN_add_word(3)>, L<BN_set_bit(3)>

=head1 COPYRIGHT

Copyright 2000-2020 The OpenSSL Project Authors. All Rights Reserved.

Licensed under the Apache License 2.0 (the "License"). You may not use
this file except in compliance with the License. You can obtain a copy
in the file LICENSE in the source distribution or at
L<https://www.openssl.org/source/license.html>.

=cut
```

1. Some edits might be required after reading the comments, before final push, 
2. I just copied the comments and tried to keep the format as close as possible to the doc/man3/BN_add.pod 
3. Not sure what to do about  it needs external reference `#define BN_GF2m_cmp(a, b) BN_ucmp((a), (b))`

**Comment by romen** [comment]
It looks very promising! Thanks @varunsood189!

How familiar are you with git and github work flows?

I can guide you on how to contribute to the openssl code base using our usual work flow, now I will give you the short version but just ask if you need more details/guidance. 

You should create a branch with this changes in your local clone of our git repository, and then push it to your github clone (fork) of openssl/openssl. 
When you do so github will prompt you to create a new PR (pull request) to the original repository, which would allow us to review the code and leave feedback, while also running our CI setup to avoid potential regressions or mistakes.

Part of this CI will also build the docs and check coherence and correctness of the new pod file.

In parallel you should submit a CLA so that we can incorporate this and your future contributions into the project! More about this administrative requirement is available at https://www.openssl.org/policies/cla.html

We look forward to have you on board among the openssl contributors!
